### PR TITLE
Draft of C API design guidelines

### PR DESCRIPTION
**Pull request by encukou**
See: https://github.com/capi-workgroup/api-evolution/issues/44

Please don't yet proofread this for grammar (or ReST syntax). Correct a typo if you see it, but focus on organization and content.

### CODE REVIEW

File: peps/pep-9999.rst [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+
+
+Applicability
+=============
+
+These guidelines only apply to API that is:
+
+* *newly added*. Existing API does not necessarily follow these guidelines;
+  this PEP is not the place for plans to replace, deprecate or remove it.
+* *pubic*: meant for users. That is, not in the *private* or *internal*
```
Comment (JelleZijlstra) [inline peps/pep-9999.rst]:
```suggestion
* *public*: meant for users. That is, not in the *private* or *internal*
```

File: peps/pep-9999.rst [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+``Py_FatalError``; specific groups of APIs use a longer prefix,
+e.g. ``PyUnicode_`` for string functions.
+Use an existing prefix when applicable. If you want to add a new prefix,
+contact the C API working group.
+
+The prefix is in mixed case, even in macro names (e.g. ``PyUnicode_AS_STRING``).
+
+Unstable API is prefixed with ``PyUnstable_`` instead of ``Py``,
+e.g. ``PyUnstable_Long_IsCompact`` or (hypothetically)
+``PyUnicode_String_GET_SIZE``.
```
Comment (JelleZijlstra) [inline peps/pep-9999.rst]:
```suggestion
``PyUnstable_String_GET_SIZE``.
```
I think this is what you meant?

File: peps/pep-9999.rst [hunk 3]
```diff
@@ -0,0 +1,820 @@
...
+- C99
+- C88 with several select C99 features:
+
+  - ``<stdint.h>`` and ``<inttypes.h>``
+  - ``static inline`` functions
+  - designated initializers
+  - intermingled declarations
+  - line comments (``//``)
+
+- C++
```
Comment (JelleZijlstra) [inline peps/pep-9999.rst]: Which version of C++?

File: peps/pep-9999.rst (hunk 2 above)
Comment (encukou) [reply peps/pep-9999.rst]: Yes, thanks!

File: peps/pep-9999.rst (hunk 3 above)
Comment (encukou) [reply peps/pep-9999.rst]: Unspecified for now. See https://github.com/capi-workgroup/api-evolution/issues/22

File: peps/pep-9999.rst [hunk 4]
```diff
@@ -643,7 +643,22 @@ a *borrowed reference*.
...
+   * “The returned reference is borrowed from p” -- the lender is crucial
+     information
+   * “that is” -- emphasizes that “borrowed from X” and “only valid as long as
+     you hold a reference to X” are equivalent
+   * “only valid as long as you hold a reference to p” -- this should be
+     spelled out rather than e.g. hidden behind a glossary link.
+   * “To get a strong reference...” -- there should always be an alternative
+     that gives you a strong ref.
+
+Make sure the documentation lists *Return value: Borrowed reference*,
```
Comment (erlend-aasland) [inline peps/pep-9999.rst:531]: Perhaps make it explicit that this annotation is added via `Doc/data/refcounts.dat`.

**Comment by encukou** [comment]
I'm now adding this to https://github.com/capi-workgroup/api-evolution, in several chunks.

**Pull request closed without merging**